structured/privilege.go

```diff
@@ -0,0 +1,258 @@
+func (p *PrivilegeDescriptor) findOrCreateUser(user string) *UserPrivileges {
+	idx := sort.Search(len(p.Users), func(i int) bool {
+		return p.Users[i].User >= user
+	})
+	if idx == len(p.Users) {
+		// Not found but should be inserted at the end.
+		p.Users = append(p.Users, &UserPrivileges{User: user})
+	} else if p.Users[idx].User == user {
+		// Found.
+	} else {
+		// New element to be inserted at i.
+		p.Users = append(p.Users, nil)
+		copy(p.Users[idx+1:], p.Users[idx:])
+		p.Users[idx] = &UserPrivileges{User: user}
+	}
+	return p.Users[idx]
+}
+
+// removeUser looks for a given user in the list and removes it if present.
+func (p *PrivilegeDescriptor) removeUser(user string) {
+	idx := sort.Search(len(p.Users), func(i int) bool {
+		return p.Users[i].User >= user
+	})
+	if idx == len(p.Users) || p.Users[idx].User != user {
+		// Not found.
+		return
+	}
+
+	copy(p.Users[idx:], p.Users[idx+1:])
+	p.Users[len(p.Users)-1] = nil
+	p.Users = p.Users[:len(p.Users)-1]
+}
+
+// DefaultDatabasePrivilegeDescriptor returns a privilege descriptor
+// with ALL privileges for the root user.
+func DefaultDatabasePrivilegeDescriptor() *PrivilegeDescriptor {
+	return &PrivilegeDescriptor{
+		Users: []*UserPrivileges{
+			{
+				User:       security.RootUser,
+				Privileges: (1 << uint32(Privilege_ALL)),
+			},
+		},
+	}
+}
+
+// Grant adds new privileges to this descriptor for a given list of users.
+// TODO(marc): if all privileges other than ALL are set, should we collapse
+// them into ALL?
+func (p *PrivilegeDescriptor) Grant(user string, privileges PrivilegeList) {
+	userPriv := p.findOrCreateUser(user)
+	for _, p := range privileges {
+		if p == Privilege_ALL {
+			// TODO(marc): even though this should not happen, error out if
+			// ALL is not the only privilege being granted.
+			userPriv.Privileges = (1 << uint32(p))
+			break
+		} else if userPriv.Privileges&(1<<uint32(Privilege_ALL)) != 0 {
+			// If we have ALL, don't set others.
+			continue
+		}
+		userPriv.Privileges |= (1 << uint32(p))
+	}
+}
+
+// Revoke removes privileges from this descriptor for a given list of users.
+func (p *PrivilegeDescriptor) Revoke(user string, privileges PrivilegeList) {
+	userPriv, ok := p.findUser(user)
+	if !ok {
+		// Removing privileges from a user without privileges is a no-op.
+		return
+	}
+	for _, p := range privileges {
+		if p == Privilege_ALL {
+			// TODO(marc): even though this should not happen, error out if
+			// ALL is not the only privilege being revoked.
+			userPriv.Privileges = 0
+			break
+		}
+		if userPriv.Privileges&(1<<uint32(Privilege_ALL)) != 0 {
+			// We're revoking something other than ALL, clear ALL
+			// and set all others.
+			userPriv.Privileges = 0
+			for _, v := range privilegeValues {
+				if v != Privilege_ALL {
+					userPriv.Privileges |= (1 << uint32(v))
+				}
+			}
+		}
+		// One doesn't see "AND NOT" very often.
+		userPriv.Privileges &^= (1 << uint32(p))
```

Comment:
yup. I've never seen "AND NOT" before. hence the comment: https://golang.org/ref/spec#Arithmetic_operators